devito/symbolics/manipulation.py

```diff
@@ -319,12 +319,152 @@ def reuse_if_untouched(expr, args, evaluate=False):
         return expr.func(*args, evaluate=evaluate)
 
 
-def evalmin(a, b):
+def evalmin(input=None, assumptions=None):
     """
-    Simplify min(a, b) if possible.
+    Simplify evalmin if possible and return nested MIN/MAX expression.
+
+    Parameters
+    ----------
+    input : a list of the symbols to be simplified
+        The candidates to be evaluated. Defaults to None.
+    assumptions : A list of assumptions formed as relationals, assumed to be True
+        Defaults to None.
+
+    Examples
+    --------
+    'Assuming no values are know for a, b, c, d but we know that d<=a and c>=b we can'
+    'safely drop `a` and `c` from the cancidate list'
+
+    >>> a = Symbol('a')
+    >>> b = Symbol('b')
+    >>> c = Symbol('c')
+    >>> d = Symbol('d')
+
+    >>> evalmin([a, b, c, d], [[Le(d, a), Ge(c, b)]])
+    MIN(b, d)
     """
+
+    if len(input) == 1:
+        return input[0]
+
+    mapper = {}
+    if assumptions:
+        for asm in assumptions:
+            if set(asm.args).issubset(input):
+                if (asm.__class__ in (Le, Lt)):
+                    mapper.update({asm.args[1]: asm.args[0]})
+                elif (asm.__class__ in (Ge, Gt)):
+                    mapper.update({asm.args[0]: asm.args[1]})
+
+    mapper = transitive_closure(mapper)
+    input = [i.subs(mapper) for i in input]
+
     try:
-        bool(min(a, b))  # Can it be evaluated or simplified?
-        return min(a, b)
+        bool(Min(*input))  # Can it be evaluated or simplified?
+        input = list(OrderedDict.fromkeys(input))
+        if len(input) == 1:
+            return input[0]
+        else:
+            input = list(input)
+            exp = MIN(input[0], input[1])
+            for i in input[2:]:
+                exp = MIN(exp, i)
+            return exp
     except TypeError:
-        return MIN(a, b)
+        exp = MIN(input[0], input[1])
+        for i in input[2:]:
+            exp = MIN(exp, i)
```

Comment:
`MIN(*input)` ....

Reply:
This will not trigger nesting...
I want to have MIN(  MIN ( exp....